libc-test/test/style.rs

```diff
@@ -0,0 +1,357 @@
+    walk(Path::new(&arg), &mut errors);
+
+    if errors.errs {
+        panic!("found some lint errors");
+    } else {
+        println!("good style!");
+    }
+}
+
+fn walk(path: &Path, err: &mut Errors) {
+    for entry in t!(path.read_dir()).map(|e| t!(e)) {
+        let path = entry.path();
+        if t!(entry.file_type()).is_dir() {
+            walk(&path, err);
+            continue;
+        }
+
+        let name = entry.file_name().into_string().unwrap();
+        match &name[..] {
+            n if !n.ends_with(".rs") => continue,
+
+            "lib.rs" | "macros.rs" => continue,
+
+            _ => {}
+        }
+
+        let mut contents = String::new();
+        t!(t!(fs::File::open(&path)).read_to_string(&mut contents));
+
+        let file = t!(syn::parse_file(&contents));
+        StyleChecker::new(|line, msg| err.error(&path, line, msg)).visit_file(&file);
+    }
+}
+
+struct Errors {
+    errs: bool,
+}
+
+struct StyleChecker<F>
+where
+    F: FnMut(usize, &str),
+{
+    state: State,
+    // FIXME: see StyleChecker::set_state
+    _s_macros: usize,
+    f_macros: usize,
+    on_err: F,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+enum State {
+    Start,
+    Imports,
+    Typedefs,
+    Structs,
+    Constants,
+    FunctionDefinitions,
+    Functions,
+    Modules,
+}
+
+/// Similar to [syn::ExprIf] except with [syn::Attribute]
+/// as the condition instead of [syn::Expr].
+struct CfgExprIf {
+    _attr: syn::Attribute,
+    then_branch: Vec<syn::Item>,
+    else_branch: Option<Box<CfgExprElse>>,
+}
+
+enum CfgExprElse {
+    Block(Vec<syn::Item>),
+    If(CfgExprIf),
+}
+
+impl<F> StyleChecker<F>
+where
+    F: FnMut(usize, &str),
+{
+    fn new(on_err: F) -> Self {
+        Self {
+            state: State::Start,
+            _s_macros: 0,
+            f_macros: 0,
+            on_err,
+        }
+    }
+
+    fn set_state(&mut self, new_state: State, line: usize) {
+        if self.state > new_state {
+            (self.on_err)(
+                line,
+                &format!(
+                    "{} found after {} when it belongs before",
+                    new_state.desc(),
+                    self.state.desc()
+                ),
+            );
+        }
+
+        if self.f_macros == 2 {
+            self.f_macros += 1;
+            (self.on_err)(line, "multiple f! macros in one module");
+        }
+
+        // FIXME(#4109): multiple should be allowed if at least one is `cfg(not) within `cfg_if`.
+        // For now just disable this and check by hand.
+        // if self.s_macros == 2 {
+        //     self.s_macros += 1;
+        //     (self.on_err)(line, "multiple s! macros in one module");
+        // }
+
+        self.state = new_state;
+    }
+
+    /// Visit the items inside the [CfgExprIf] and return the final
+    /// state after, which is conservatively the minimum of all branches
+    /// since the branches could have diverged in state.
```

Comment:
Not sure what the best way to handle the state while branching is, the old `style.rs` script didn't consider branching at all but only got away with it here because it thought it was in an `impl` due to this comment:

https://github.com/rust-lang/libc/blob/48692735ab221c30e63e413ede29272b406dfd91/src/fixed_width_ints.rs#L45

I think the two options are to keep the same behavior as the old script which would mean this `cfg_if!` block would have to be separated into two (duplicating the `#[cfg(...)]` or to do some sort of join with the min or max of the states of all the branches.

Reply:
Would it be possible to treat each arm as if each item has the `#[cfg(...)]` (or `#[cfg(not(...))]` for the default directly applied?

I think if you save the state before entering each `cfg_if` branch and then restore it after, things will probably work out okay.

We might want to adjust this behavior later so as long as existing files pass, I'm not too concerned.

Reply:
Also in general, I'm not really concerned about whether configuration happens via `cfg_if` or `cfg`. The original checker was pretty strict about this, but IMO I'd rather see `#[cfg(...)]` if it's a one-time thing.

The more important thing is have it yell at you and suggest `cfg_if` if the same `cfg` is repeated, either in multiple `#[cfg(...)]`s or in multiple `cfg_if`s. But that's something that could be added later.

Reply:
> Would it be possible to treat each arm as if each item has the #[cfg(...)] (or #[cfg(not(...))] for the default directly applied?

Sorry, what do you mean by this, what is the default that needs to be directly applied?

> I think if you save the state before entering each `cfg_if` branch and then restore it after, things will probably work out okay.
> 
> We might want to adjust this behavior later so as long as existing files pass, I'm not too concerned.

Yeah makes sense, let's do that for now

> The more important thing is have it yell at you and suggest `cfg_if` if the same `cfg` is repeated, either in multiple `#[cfg(...)]`s or in multiple `cfg_if`s. But that's something that could be added later.

Sounds good, I'd be happy to help out with adding some of these new rules that are easier with `syn` but I just want to work on getting the basics down first :)

Reply:
> > Would it be possible to treat each arm as if each item has the #[cfg(...)] (or #[cfg(not(...))] for the default directly applied?
> 
> Sorry, what do you mean by this, what is the default that needs to be directly applied?

I missed a `)`, it should have been `(or #[cfg(not(...))] for the default)`referring to the last `else` condition. But I'm realizing this doesn't make a whole lot of sense anyway, so just ignore that comment for now.

> > The more important thing is have it yell at you and suggest `cfg_if` if the same `cfg` is repeated, either in multiple `#[cfg(...)]`s or in multiple `cfg_if`s. But that's something that could be added later.
> 
> Sounds good, I'd be happy to help out with adding some of these new rules that are easier with `syn` but I just want to work on getting the basics down first :)

This is much appreciated :) the tooling here has been in need of a refresh for a while!